all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,526 @@
+
+When a user creates an API key, the completion screen can provide a download of the config file that will have the
+address, namespace, and API key set. So it can say something like:
+
+```
+API key is created, make sure you copy it here because it is not stored:
+
+    <some long api key string>
+
+Alternatively, you can download it as a configuration file:
+
+    <some button or something that downloads a file name temporalio-client.toml>
+
+You can place this configuration file in the default place used by CLIs/SDKs:
+
+* macOS - `$HOME/Library/Application Support/temporalio/temporal-client.toml`
+* Windows - `%AppData%/temporalio/temporal-client.toml` (often `C:\Users\<myuser>\AppData\Roaming\temporalio\temporal-client.toml`)
+* Linux - `$HOME/.config/temporalio/temporal-client.toml`
+
+This will be automatically consumed by the SDKs/CLI. This file can also be put anywhere else and `TEMPORAL_CONFIG_FILE`
+environment variable can be set to point to it.
+```
+
+#### Possible CLI Creation Flow
+
+Assuming cloud support is present in `temporal` CLI, you can have the normal certificate generation steps, e.g.
+
+```
+temporal cloud gen ca ...
+temporal cloud gen leaf ...
+temporal cloud namespace add-accepted-client-ca ...
+```
+
+And that last step can by default set a `temporal-cloud` profile in the configuration with the address, namespace, and
+mTLS certificates. So now all you have to do to use cloud in CLI is either `--profile temporal-cloud` or set
+`TEMPORAL_PROFILE=temporal-cloud` environment variable. Granted you may want better certificate management than that and
+whether this is done by default can be discussed, but the idea is the same.
+
+#### Multi-Namespace Config Use
+
+Some users may want the same configuration for multiple namespaces (or multiple addresses or multiple API keys or
+whatever). Users can supply partial configuration. For instance, you can have this config:
+
+```toml
+[profile.temporal-cloud]
+address = "my-dev-ns.a1b2c.tmprl.cloud:7233"
+tls.client_cert_path = "path/to/my/dev-cert.pem"
+tls.client_key_path = "path/to/my/dev-cert.pem"
+```
+
+And then still use CLI like
+
+    temporal workflow list --namespace my-specific-namespace
+
+This will use the config file information for all but the namespace. Similarly you could just set the
+`TEMPORAL_NAMESPACE` env var to `my-specific-namespace` and run:
+
+    temporal workflow list
+
+Similarly, if you're using an SDK, you can load the config and just switch namespaces in code. So you can have:
+
+```python
+config = envconfig.load_client_connect_config()
+client = await Client.connect(namespace="my-specific-namespace", **config)
+```
+
+## Specification
+
+### Values and File Format
+
+* The format is a TOML object with the following fields:
+  * `profile` - Object with keys as profile names. The default profile should be named `default`.
+    * `<name>` - Profile object with configuration values. Profile names are case-insensitive. 💭 Why? They are needed
```

Comment:
No longer true:
~They are needed for environment variables which are case insensitive~

Still, I think case-insensitive profile names make sense. Could want to reopen the env variable avenue later, and that would be a likely source of error for users anyway.

Reply:
> Still, I think case-insensitive profile names make sense

I'm not really looking to protect from typos here. I don't think case sensitivity of profile name is any more painful than case sensitivity of config keys (we require all lowercase) or env var names (we require all upper) or CLI flag names or whatever. If we need to relax case sensitivity later we can, but I think it's normal for people to expect it.

> Are we ok accepting any valid TOML key as profile name? even quoted keys?

I am ok with this for the most part, or at least I can't see any obvious reason why not. But I admittedly did not cover advanced validation of things in this proposal. I tend to want to accept loosely and have the usage fail if needed as opposed to have lots of matching validation in lots of different languages.